OAK-10896: - Add OSGI configuration variable for removal of deleted properties and orphaned nodes

shodaaan opened the pull request:
- create class FullGCOptions which will hold the different parameters used for running full GC
- added variables to Configuration.java for 2 full GC modes: GAP_ORPHANS and EMPTY_PROPERTIES
- updated VersionGarbageCollector constructor, as well as DocumentNodeStore Builder / Service and DocumentNodeStore to get values received from Configuration via the FullGCOptions class

- applied the new full GC mode options in VersionGarbageCollector via the fullGCOptions parameter passed in the constructor
- added unit test methods for testing feature toggle and configuration options set in DocumentNodeStoreBuilder, based on existing testing of feature toggles and configuration settings in unit tests
- fixed CommitBuilderTest test failures by changing expected exception type

shodaaan committed:
- create class FullGCOptions which will hold the different parameters used for running full GC
- added variables to Configuration.java for 2 full GC modes: GAP_ORPHANS and EMPTY_PROPERTIES
- udpated VersionGarbageCollector constructor, as well as DocumentNodeStore Builder / Service and DocumentNodeStore to get values received from Configuration via the FullGCOptions class

shodaaan committed:
- applied the new full GC mode options in VersionGarbageCollector via the fullGCOptions parameter passed in the constructor
- added unit test methods for testing feature toggle and configuration options set in DocumentNodeStoreBuilder, based on existing testing of feature toggles and configuration settings in unit tests
- fixed CommitBuilderTest test failures by changing expected exception type

stefan-egli reviewed (commented):
I think the main aspect of this work is the question whether we want actual feature (toggles) for the new config options or not (despite the jira title). In my opinion we wouldn't automatically need a full-fledged feature toggle for each new config we introduce, as they probably have a higher cost. With "gc mode" related ones, I'd vote to treat them as plain osgi configuration only.

Another aspect is the question of having a separate `FullGCOptions` class or merging it with the existing `VersionGCOptions`. We had a history of treating things separate (eg `VersionGarbageCollector.FullGC`) or combined (`VersionGCRecommendations`) - so we weren't consistent. For `FullGCOptions` case I'm not sure there is advantages of having it separate though? One argument that comes to mind for combining is that full-gc already relies on the existing options, so from that point of view it doesn't live separate. This is just my opinion though...

PS: please avoid wildcard imports

rishabhdaim commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java:
```diff
@@ -38,22 +38,11 @@
-import org.apache.jackrabbit.oak.plugins.document.RevisionContextWrapper;
-import org.apache.jackrabbit.oak.plugins.document.SweepHelper;
-import org.apache.jackrabbit.oak.plugins.document.VersionGCSupport;
+import org.apache.jackrabbit.oak.plugins.document.*;
```
Please avoid `*` imports.

shodaaan replied:
fixed

rishabhdaim commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java:
```diff
@@ -69,11 +58,7 @@
-import static org.apache.jackrabbit.oak.plugins.document.util.Utils.isFullGCEnabled;
-import static org.apache.jackrabbit.oak.plugins.document.util.Utils.isEmbeddedVerificationEnabled;
-import static org.apache.jackrabbit.oak.plugins.document.util.Utils.timestampToString;
+import static org.apache.jackrabbit.oak.plugins.document.util.Utils.*;
```
Please avoid `*` imports.

stefan-egli replied:
+1

shodaaan replied:
fixed

rishabhdaim commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java:
```diff
@@ -34,9 +34,7 @@
-import static org.apache.jackrabbit.oak.plugins.document.DocumentNodeStoreService.DEFAULT_FULL_GC_ENABLED;
-import static org.apache.jackrabbit.oak.plugins.document.DocumentNodeStoreService.DEFAULT_EMBEDDED_VERIFICATION_ENABLED;
-import static org.apache.jackrabbit.oak.plugins.document.DocumentNodeStoreService.DEFAULT_THROTTLING_ENABLED;
+import static org.apache.jackrabbit.oak.plugins.document.DocumentNodeStoreService.*;
```
Please avoid `*` imports.

shodaaan replied:
fixed

rishabhdaim commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStore.java:
```diff
@@ -661,7 +655,9 @@ public DocumentNodeStore(DocumentNodeStoreBuilder<?> builder) {
-                this, builder.createVersionGCSupport(), isFullGCEnabled(builder), false, isEmbeddedVerificationEnabled(builder));
+                this, builder.createVersionGCSupport(), isFullGCEnabled(builder), false,
+                isEmbeddedVerificationEnabled(builder),
+                new FullGCOptions(isFullGCModeGapOrphansEnabled(builder), Utils.isFullGCModeEmptyPropertiesEnabled(builder)));
```
```suggestion
                new FullGCOptions(isFullGCModeGapOrphansEnabled(builder), isFullGCModeEmptyPropertiesEnabled(builder)));
```

shodaaan replied:
fixed

rishabhdaim commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBuilder.java:
```diff
@@ -312,6 +316,20 @@ public boolean isEmbeddedVerificationEnabled() {
+        return thisBuilder();
+    }
+
+    public boolean isFullGCModeGapOrphansEnabled() { return this.fullGCModeGapOrphansEnabled; }
```
```suggestion
    public boolean isFullGCModeGapOrphansEnabled() { 
    return this.fullGCModeGapOrphansEnabled; 
    }
```

shodaaan replied:
fixed

rishabhdaim commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBuilder.java:
```diff
@@ -312,6 +316,20 @@ public boolean isEmbeddedVerificationEnabled() {
+        return thisBuilder();
+    }
+
+    public boolean isFullGCModeEmptyPropertiesEnabled() { return this.fullGCModeEmptyPropertiesEnabled; }
```
```suggestion
    public boolean isFullGCModeEmptyPropertiesEnabled() { 
    return this.fullGCModeEmptyPropertiesEnabled; 
    }
```

rishabhdaim commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBuilder.java:
```diff
@@ -380,6 +398,22 @@ public Feature getDocStoreEmbeddedVerificationFeature() {
+    }
+
+    @Nullable
+    public Feature getDocStoreFullGCModeGapOrphansFeature() { return this.docStoreFullGCModeGapOrphansFeature;    }
```
```suggestion
    public Feature getDocStoreFullGCModeGapOrphansFeature() { 
    return this.docStoreFullGCModeGapOrphansFeature;    
    }
```

rishabhdaim commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBuilder.java:
```diff
@@ -380,6 +398,22 @@ public Feature getDocStoreEmbeddedVerificationFeature() {
+    }
+
+    @Nullable
+    public Feature getDocStoreFullGCModeEmptyPropertiesFeature() { return this.docStoreFullGCModeEmptyPropertiesFeature; };
```
```suggestion
    public Feature getDocStoreFullGCModeEmptyPropertiesFeature() { 
    return this.docStoreFullGCModeEmptyPropertiesFeature; 
    }
```

rishabhdaim commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreService.java:
```diff
@@ -199,6 +201,16 @@ public class DocumentNodeStoreService {
+    /**
+     * Feature toggle name to enable full GC mode GAP_ORPHANS cleanup for Mongo Document Store
+     */
+    private static final String FT_NAME_FULL_GC_MODE_GAP_ORPHANS = "FT_FULL_GC_MODE_GAP_ORPHANS";
```
please add the OAK ticket id in the feature toggle name.

rishabhdaim commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreService.java:
```diff
@@ -199,6 +201,16 @@ public class DocumentNodeStoreService {
+    /**
+     * Feature toggle name to enable full GC mode EMPTY_PROPERTIES cleanup for Mongo Document Store
+     */
+    private static final String FT_NAME_FULL_GC_MODE_EMPTY_PROPERTIES = "FT_FULL_GC_MODE_EMPTY_PROPERTIES";
```
same as above

rishabhdaim commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/rdb/RDBDocumentNodeStoreBuilder.java:
```diff
@@ -139,6 +163,25 @@ public Feature getDocStoreFullGCFeature() {
+
+    @Override
+    @Nullable
+    public Feature getDocStoreFullGCModeGapOrphansFeature() { return null; }
```
```suggestion
    public Feature getDocStoreFullGCModeGapOrphansFeature() { 
    return null; 
    }
```

shodaaan replied:
fixed

rishabhdaim commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/rdb/RDBDocumentNodeStoreBuilder.java:
```diff
@@ -139,6 +163,25 @@ public Feature getDocStoreFullGCFeature() {
+
+    @Override
+    @Nullable
+    public Feature getDocStoreFullGCModeEmptyPropertiesFeature() { return null; };
```
```suggestion
    public Feature getDocStoreFullGCModeEmptyPropertiesFeature() { 
    return null; 
    }
```

shodaaan replied:
fixed

rishabhdaim commented on oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/rdb/RDBDocumentNodeStoreBuilderTest.java:
```diff
@@ -16,9 +16,7 @@
-import static org.junit.Assert.assertFalse;
-import static org.junit.Assert.assertNull;
-import static org.junit.Assert.fail;
+import static org.junit.Assert.*;
```
please avoid `*` imports.

shodaaan replied:
fixed

rishabhdaim commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/util/Utils.java:
```diff
@@ -1009,6 +1009,28 @@ public static boolean isEmbeddedVerificationEnabled(final DocumentNodeStoreBuild
+     * @param builder instance for DocumentNodeStoreBuilder
+     * @return true if full GC is enabled else false
+     */
+    public static boolean isFullGCModeGapOrphansEnabled(final DocumentNodeStoreBuilder<?> builder) {
```
please add unit cases for this method.

shodaaan replied:
no longer necessary

rishabhdaim commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/util/Utils.java:
```diff
@@ -1009,6 +1009,28 @@ public static boolean isEmbeddedVerificationEnabled(final DocumentNodeStoreBuild
+     * @param builder instance for DocumentNodeStoreBuilder
+     * @return true if full GC is enabled else false
+     */
+    public static boolean isFullGCModeEmptyPropertiesEnabled(final DocumentNodeStoreBuilder<?> builder) {
```
please add unit cases for this method.

shodaaan replied:
no longer necessary

rishabhdaim commented:
> I think the main aspect of this work is the question whether we want actual feature (toggles) for the new config options or not (despite the jira title). In my opinion we wouldn't automatically need a full-fledged feature toggle for each new config we introduce, as they probably have a higher cost. With "gc mode" related ones, I'd vote to treat them as plain osgi configuration only.
> 
> Another aspect is the question of having a separate `FullGCOptions` class or merging it with the existing `VersionGCOptions`. We had a history of treating things separate (eg `VersionGarbageCollector.FullGC`) or combined (`VersionGCRecommendations`) - so we weren't consistent. For `FullGCOptions` case I'm not sure there is advantages of having it separate though? One argument that comes to mind for combining is that full-gc already relies on the existing options, so from that point of view it doesn't live separate. This is just my opinion though...
> 
> PS: please avoid wildcard imports

I agree with @stefan-egli that we don't need feature toggles for each FULL_GC_MODE, just one feature toggle is enough to enable/disable FULL_GC.

However, I recommend using an OSGI config (with values limited from an enum of FULL_GC_MODE) to select the FULL_GC_MODE.

shodaaan committed:
- created new OSGI configuration fullGCMode that is passed to the VersionGarbageCollector constructor via DocumentNodeStoreBuilder
- added method to VersionGarbageCollector for setting the private static fullGCMode variable in the constructor based on the OSGI configuration value and added unit tests to the VersionGCTest class for the logic
- created unit tests for the setting of the new configuration variable and for using it in the DocumentNodeStore classes (Builder, Service, etc.)
- adjusted unit tests in VersionGarbageCollector by setting the private static fullGCMode variable to the desired value after creating a secondary node store
- removed previously created config variables fullGCModeGapOrphans / EmptyProperties and their usage in DocumentNodeStore classes. Also removed previously created class FullGCOptions which was used to pass these variables to VersionGarbageCollector

stefan-egli commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStore.java:
```diff
@@ -35,13 +35,7 @@
-import static org.apache.jackrabbit.oak.plugins.document.util.Utils.isEmbeddedVerificationEnabled;
-import static org.apache.jackrabbit.oak.plugins.document.util.Utils.isThrottlingEnabled;
-import static org.apache.jackrabbit.oak.plugins.document.util.Utils.pathToId;
+import static org.apache.jackrabbit.oak.plugins.document.util.Utils.*;
```
no wildcard imports

shodaaan replied:
fixed

stefan-egli commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreService.java:
```diff
@@ -140,6 +140,8 @@ public class DocumentNodeStoreService {
     static final boolean DEFAULT_FULL_GC_ENABLED = false;
     static final boolean DEFAULT_EMBEDDED_VERIFICATION_ENABLED = true;
+    static final int DEFAULT_FULL_GC_MODE = 0;
+    static final boolean DEFAULT_FULL_GC_EMPTY_PROPERTIES_ENABLED = false;
```
this ls unused

shodaaan replied:
fixed.

stefan-egli commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreService.java:
```diff
@@ -199,6 +201,16 @@ public class DocumentNodeStoreService {
      */
     private static final String FT_NAME_EMBEDDED_VERIFICATION = "FT_EMBEDDED_VERIFICATION_OAK-10633";
 
+    /**
+     * Feature toggle name to enable full GC mode GAP_ORPHANS cleanup for Mongo Document Store
+     */
+    private static final String FT_NAME_FULL_GC_MODE_GAP_ORPHANS = "FT_FULL_GC_MODE_GAP_ORPHANS_OAK-10896";
+
+    /**
+     * Feature toggle name to enable full GC mode EMPTY_PROPERTIES cleanup for Mongo Document Store
+     */
+    private static final String FT_NAME_FULL_GC_MODE_EMPTY_PROPERTIES = "FT_FULL_GC_MODE_EMPTY_PROPERTIES_OAK-10896";
+
```
I was under the impression we decided not to use feature toggles, in which case this (and related code) is not necessary

shodaaan replied:
fixed, thank you for the comment.

stefan-egli commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreService.java:
```diff
@@ -276,6 +290,8 @@ protected void activate(ComponentContext context, Configuration config) throws E
         cancelInvalidationFeature = Feature.newFeature(FT_NAME_CANCEL_INVALIDATION, whiteboard);
         docStoreFullGCFeature = Feature.newFeature(FT_NAME_FULL_GC, whiteboard);
         docStoreEmbeddedVerificationFeature = Feature.newFeature(FT_NAME_EMBEDDED_VERIFICATION, whiteboard);
+        docStoreFullGCModeGapOrphansFeature = Feature.newFeature(FT_NAME_FULL_GC_MODE_GAP_ORPHANS, whiteboard);
+        docStoreFullGCModeEmptyPropertiesFeature = Feature.newFeature(FT_NAME_FULL_GC_MODE_EMPTY_PROPERTIES, whiteboard);
```
(same as above comment - I'll not repeat this comment further though)

shodaaan replied:
fixed, thank you for the comment.

stefan-egli commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java:
```diff
@@ -1092,61 +1124,63 @@ public void collectGarbage(final NodeDocument doc, final GCPhases phases) {
             }
 
             if (!isDeletedOrOrphanedNode(traversedState, greatestExistingAncestorOrSelf, phases, doc)) {
+
```
please avoid inserting empty lines unless say demarking new code for readability etc

stefan-egli commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java:
```diff
@@ -1092,61 +1124,63 @@ public void collectGarbage(final NodeDocument doc, final GCPhases phases) {
             }
 
             if (!isDeletedOrOrphanedNode(traversedState, greatestExistingAncestorOrSelf, phases, doc)) {
+
                 // here the node is not orphaned which means that we can reach the node from root
-                switch(fullGcMode) {
-                    case NONE : {
+                switch (fullGcMode) {
+                    case NONE: {
```
please avoid unrelated reformatting

stefan-egli replied:
(same applies to other case statements below)

shodaaan replied:
Actually, this is the suggested way to format switch statements, in fact it was automatically suggested by the IntelliJ Java text editor.
Also, other switch statements in the Oak project use this formatting, with no space between the value and ":".

stefan-egli replied:
still, this in unrelated and we should not increase PR change surface when that is not necessary or part of the PR

stefan-egli commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java:
```diff
@@ -1092,61 +1124,63 @@ public void collectGarbage(final NodeDocument doc, final GCPhases phases) {
-                    case ORPHANS_EMPTYPROPS_UNMERGED_BC : {
+                    case ORPHANS_EMPTYPROPS_UNMERGED_BC: {
                         collectDeletedProperties(doc, phases, op, traversedState);
                         collectUnmergedBranchCommits(doc, phases, op, toModifiedMs);
                         break;
                     }
-                    case ORPHANS_EMPTYPROPS_BETWEEN_CHECKPOINTS_WITH_UNMERGED_BC : {
+                    case ORPHANS_EMPTYPROPS_BETWEEN_CHECKPOINTS_WITH_UNMERGED_BC: {
                         collectDeletedProperties(doc, phases, op, traversedState);
                         collectUnmergedBranchCommits(doc, phases, op, toModifiedMs);
                         collectRevisionsOlderThan24hAndBetweenCheckpoints(doc, toModifiedMs, phases, op);
                         break;
                     }
-                    case ORPHANS_EMPTYPROPS_BETWEEN_CHECKPOINTS_NO_UNMERGED_BC : {
+                    case ORPHANS_EMPTYPROPS_BETWEEN_CHECKPOINTS_NO_UNMERGED_BC: {
                         collectDeletedProperties(doc, phases, op, traversedState);
                         collectRevisionsOlderThan24hAndBetweenCheckpoints(doc, toModifiedMs, phases, op);
                         break;
                     }
                 }
-                // only add if there are changes for this doc
-                if (op.hasChanges()) {
-                    op.equals(MODIFIED_IN_SECS, doc.getModified());
-                    garbageDocsCount++;
-                    totalGarbageDocsCount++;
-                    monitor.info("Collected [{}] garbage count in [{}]", op.getChanges().size(), doc.getId());
-                    AUDIT_LOG.info("<Collected> [{}] garbage count  in [{}]", op.getChanges().size(), doc.getId());
-                    updateOpList.add(op);
-                }
             }
+            // only add if there are changes for this doc
+            if (op.hasChanges()) {
+                op.equals(MODIFIED_IN_SECS, doc.getModified());
+                garbageDocsCount++;
+                totalGarbageDocsCount++;
+                monitor.info("Collected [{}] garbage count in [{}]", op.getChanges().size(), doc.getId());
+                AUDIT_LOG.info("<Collected> [{}] garbage count  in [{}]", op.getChanges().size(), doc.getId());
+                updateOpList.add(op);
+            }
+
```
why was this moved? doesn't look correct

stefan-egli commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java:
```diff
@@ -1226,7 +1260,10 @@ private boolean isDeletedOrOrphanedNode(final NodeState traversedState, final Pa
+
+            // first check if gapOrphansModeEnabled is set in fullGCOptions, then check in fullGCMode enum
             if (fullGcMode == GAP_ORPHANS || fullGcMode == GAP_ORPHANS_EMPTYPROPS) {
+
```
(as mentioned above, we should only include related changes, this new empty line doesn't seem necessary)

shodaaan replied:
fixed, thank you for the comment

stefan-egli commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java:
```diff
@@ -1226,7 +1260,10 @@ private boolean isDeletedOrOrphanedNode(final NodeState traversedState, final Pa
                 phases.stop(GCPhase.FULL_GC_COLLECT_ORPHAN_NODES);
                 return false;
             }
+
+            // first check if gapOrphansModeEnabled is set in fullGCOptions, then check in fullGCMode enum
```
not sure I see the purpose of this comment, would suggest to remove

shodaaan replied:
fixed, thank you for the comment

stefan-egli replied:
but the comment is still here?

stefan-egli commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/rdb/RDBDocumentNodeStoreBuilder.java:
```diff
@@ -128,6 +128,18 @@ public RDBDocumentNodeStoreBuilder setFullGCEnabled(boolean b) {
 
+    @Override
+    public RDBDocumentNodeStoreBuilder setFullGCMode(int v) {
+        // fullGC modes are not supported for RDB
```
should we then throw something like an `IllegalArgumentException` here?

rishabhdaim replied:
Since `fullGC` is supported for RDB, logging a `WARN` message should be sufficient.

shodaaan replied:
added log.warn, thank you for the suggestion.

stefan-egli commented on oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/CommitBuilderTest.java:
```diff
@@ -83,8 +83,8 @@ public void buildWithNullRevision() {
             builder.build(null);
-            expectNPE();
-        } catch (NullPointerException e) {
+            expectIllegalArgumentException();
+        } catch (IllegalArgumentException e) {
```
looks wrong, please revert

shodaaan replied:
The tests fail because an IllegalArgumentException is actually thrown, so I have modified them to pass.

stefan-egli replied:
it fails with a NPE for me though

stefan-egli commented on oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/CommitBuilderTest.java:
```diff
@@ -119,8 +119,8 @@ public void addNodePathNull() {
             builder.addNode((Path) null);
-            expectNPE();
-        } catch (NullPointerException e) {
+            expectIllegalArgumentException();
+        } catch (IllegalArgumentException e) {
```
same

shodaaan replied:
The tests fail because an IllegalArgumentException is actually thrown, so I have modified them to pass.

stefan-egli commented on oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/CommitBuilderTest.java:
```diff
@@ -302,6 +302,10 @@ public void defaultStartRevisions() {
         assertEquals(new RevisionVector(), c.getStartRevisions());
     }
 
+    private static void expectIllegalArgumentException() {
+        fail("IllegalArgumentException expected");
+    }
+
     private static void expectNPE() {
```
I didn't check all cases but the first I checked was still a NPE - why was this change applied?

shodaaan replied:
The tests fail because an IllegalArgumentException is actually thrown, so I have modified them to pass.

stefan-egli commented on oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollectorIT.java:
```diff
@@ -308,6 +310,9 @@ private void createSecondaryStore(LeaseCheckMode leaseCheckNode, boolean withFai
                 .setLeaseCheckMode(leaseCheckNode)
                 .setDocumentStore(ds2).setAsyncDelay(0);
         store2 = documentMKBuilder2.getNodeStore();
+        // custom fullGcMode needs to be set after each node store creation, since the VersionGarbageCollector
+        // constructor sets the fullGcMode to the value read from OSGI Configuration - method setFullGcMode
+        writeStaticField(VersionGarbageCollector.class, "fullGcMode", fullGcMode, true);
```
can't we use `setFullGcMode` here?

stefan-egli replied:
ok, I see, we had it like this previously .. I'd still vote to introducing a dedicated testing-only setter (`static void setFullGcMode(FullGCMode)`) instead of reflection (for this and a few other similar cases for fullgcmode in this and the other test class) - makes code maintenance easier imv

shodaaan replied:
I propose we make that change when we do a larger scale refactoring. Changing the way we set fullGCMode in unit tests now would also require updating all of the related unit tests (for consistency), and there are a lot of them.

We should do refactoring in a separate, dedicated ticket, not as a part of this feature, IMHO.

stefan-egli commented on oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/rdb/RDBDocumentNodeStoreBuilderTest.java:
```diff
@@ -16,9 +16,10 @@
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.fail;
 import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertNull;
-import static org.junit.Assert.fail;
```
please avoid unnecessary import reorgs

stefan-egli commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java:
```diff
@@ -211,11 +211,31 @@ static FullGCMode getFullGcMode() {
+            case 3:
+                fullGcMode = GAP_ORPHANS_EMPTYPROPS;
+                break;
+        }
+    }
```
I would suggest to add a `default` catch all section here that logs the fact that a not support mode was used and that the default mode is used instead

shodaaan replied:
Done, thank you for the suggestion.

stefan-egli commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java:
```diff
@@ -211,11 +211,31 @@ static FullGCMode getFullGcMode() {
     private final boolean fullGCEnabled;
     private final boolean isFullGCDryRun;
     private final boolean embeddedVerification;
+    private final int fullGCModeConfigValue;
```
this seems unused

shodaaan replied:
removed, thank you for the observation.

stefan-egli commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java:
```diff
@@ -228,13 +248,25 @@ static FullGCMode getFullGcMode() {
                             final boolean fullGCEnabled,
                             final boolean isFullGCDryRun,
                             final boolean embeddedVerification) {
+        this(nodeStore, gcSupport, fullGCEnabled, isFullGCDryRun, embeddedVerification, 0);
```
please use `Configuration.DEFAULT_FULL_GC_MODE` instead of 0

shodaaan replied:
fixed, thank you for the comment. I used DocumentNodeStoreService.DEFAULT_FULL_GC_MODE

stefan-egli commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java:
```diff
@@ -228,13 +248,25 @@ static FullGCMode getFullGcMode() {
+                            final boolean fullGCEnabled,
+                            final boolean isFullGCDryRun,
+                            final boolean embeddedVerification,
+                            final int fullGCModeConfigVal) {
```
what about using a shorter name like below?

```suggestion
                            final int fullGCMode) {
```

stefan-egli replied:
(same applies to different places)

stefan-egli commented on oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/VersionGCTest.java:
```diff
@@ -531,6 +531,51 @@ public void testResetFullGCDryRunMode() throws Exception {
+
+    @Test
+    public void testVersionGCLoadGCModeConfigurationNotApplicable() throws Exception {
+        int FULL_GC_MODE_NOT_ALLOWED_VALUE = 5;
```
variable names shouldn't be all capital

shodaaan replied:
Fixed, thank you for the comment

stefan-egli commented on oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/VersionGCTest.java:
```diff
@@ -531,6 +531,51 @@ public void testResetFullGCDryRunMode() throws Exception {
+                ns, new VersionGCSupport(store), true, false, false,
+                FULL_GC_MODE_NOT_ALLOWED_VALUE);
+
+        assertEquals("Starting VersionGarbageCollector with not applicable / not allowed value" +
+                "will set fullGcMode to default NONE", gc.getFullGcMode(), VersionGarbageCollector.FullGCMode.NONE);
```
this test should actually have failed,a s the coded default `VersionGarbageCollector.FullGCMode` is `GAP_ORPHANS_EMPTYPROPS` - the test only succeeds as a side-effect of `setUp` building a `DocumentNodeStore` which uses `DocumentNodeStoreBuilder.getFullGCMode` which uses an uninitialized thus default value for `fullGCMode`. 

I' suggest to change the test to first set a valid mode (eg 2) an then set an invalid mode (eg the 5) and then it should have been set to 0

shodaaan replied:
I updated the test, thank you for the suggestion.

stefan-egli commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBuilder.java:
```diff
@@ -171,6 +171,7 @@ public class DocumentNodeStoreBuilder<T extends DocumentNodeStoreBuilder<T>> {
     private boolean throttlingEnabled;
     private boolean fullGCEnabled;
     private boolean embeddedVerificationEnabled = DocumentNodeStoreService.DEFAULT_EMBEDDED_VERIFICATION_ENABLED;
+    private int fullGCMode;
```
please initialize with the default

stefan-egli replied:
(I see that we're not consistent with this .. eg as with the enabled booleans above, which default to false and that happens to be what is also the default ... perhaps we can ignore that fact plus given this is an `int`, it might be good to set it explicitly)

shodaaan replied:
I set it to the default value, thank you for the suggestion.

rishabhdaim commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java:
```diff
@@ -35,8 +35,9 @@
 import static org.apache.jackrabbit.oak.plugins.document.DocumentNodeStoreBuilder.DEFAULT_PREV_DOC_CACHE_PERCENTAGE;
 import static org.apache.jackrabbit.oak.plugins.document.DocumentNodeStoreBuilder.DEFAULT_UPDATE_LIMIT;
 import static org.apache.jackrabbit.oak.plugins.document.DocumentNodeStoreService.DEFAULT_FULL_GC_ENABLED;
-import static org.apache.jackrabbit.oak.plugins.document.DocumentNodeStoreService.DEFAULT_EMBEDDED_VERIFICATION_ENABLED;
```
please revert this un-intended change.

rishabhdaim commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java:
```diff
@@ -324,4 +325,12 @@
+
+    @AttributeDefinition(
+            name = "Document Node Store Full GC Mode",
+            description = "Boolean value indicating which Full GC mode should be enabled for " +
```
```suggestion
            description = "Integer value indicating which Full GC mode should be used for " +
```

rishabhdaim commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java:
```diff
@@ -324,4 +325,12 @@
+    @AttributeDefinition(
+            name = "Document Node Store Full GC Mode",
+            description = "Boolean value indicating which Full GC mode should be enabled for " +
+                    "document node store or not. The Default value is " + DEFAULT_FULL_GC_MODE +
```
```suggestion
                    "document node store. The Default value is " + DEFAULT_FULL_GC_MODE +
```

rishabhdaim commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java:
```diff
@@ -324,4 +325,12 @@
                     "property 'oak.documentstore.embeddedVerificationEnabled'")
     boolean embeddedVerificationEnabled() default DEFAULT_EMBEDDED_VERIFICATION_ENABLED;
+
+    @AttributeDefinition(
```
please mention that `fullGC` can be entirely disabled via `fullGCEnabled`.

shodaaan replied:
Done, thank you for the suggestion.

rishabhdaim commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java:
```diff
@@ -211,11 +211,31 @@ static FullGCMode getFullGcMode() {
+     * The configuration value will be ignored if it is set to any other values than the supported ones.
+     * @param fullGcModeConfig
+     */
+    static void setFullGcMode(int fullGcModeConfig) {
```
We decided to put this logic inside the `FullGCMode` enum.
`FullGCMode` enum should have a constructor with both `int` & `string` value and int value should match with the OSGI config.

shodaaan replied:
I think this is outside the scope of this ticket.
Doing this change now would also require refactoring the unit tests that use fullGCMode. 
I suggest that we do that in a separate refactoring ticket, since there are a lot of tests involved and we may also consider other changes to the unit tests when we do the refactoring.

shodaaan committed:
- made requested PR changes
- modified setFullGCMode method to set fullGCMode to default value if a value outside of the approved values is received through OSGI configuration
- added log warn to RDBDocumentNodeStoreBuilder

stefan-egli commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStore.java:
```diff
@@ -692,7 +694,7 @@ public int getMemory() {
         diffCache = builder.getDiffCache(this.clusterId);
 
         // check if root node exists
-        NodeDocument rootDoc = store.find(NODES, Utils.getIdFromPath(ROOT));
+        NodeDocument rootDoc = store.find(NODES, getIdFromPath(ROOT));
```
I think we should try to avoid making unrelated changes like these, they make the PR change surface unnecessarily larger (and in this particular case I don't think it is an improvement - but I guess that is a controversial topic)

shodaaan replied:
This is consistent with how we are using static method calls. The consistent approach would be IMHO to actually do this everywhere, but I think a refactoring ticket is the best place for that.

stefan-egli replied:
but it's in a random unrelated code - which is against our goal to not do unrelated changes

shodaaan replied:
Fixed.

stefan-egli commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStore.java:
```diff
@@ -716,7 +718,7 @@ public int getMemory() {
             setRoot(head);
             // make sure _lastRev is written back to store
             backgroundWrite();
-            rootDoc = store.find(NODES, Utils.getIdFromPath(ROOT));
+            rootDoc = store.find(NODES, getIdFromPath(ROOT));
```
(same as above)

shodaaan replied:
This is consistent with how we are using static method calls. The consistent approach would be IMHO to actually do this everywhere, but I think a refactoring ticket is the best place for that.

stefan-egli commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStore.java:
```diff
@@ -2603,7 +2605,7 @@ private void cleanOrphanedBranches() {
 
     private void cleanRootCollisions() {
-        String id = Utils.getIdFromPath(ROOT);
+        String id = getIdFromPath(ROOT);
```
(same as above throughout this class)

shodaaan replied:
Fixed.

stefan-egli commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreService.java:
```diff
@@ -238,6 +239,8 @@ static DocumentStoreType fromString(String type) {
     private Feature cancelInvalidationFeature;
     private Feature docStoreFullGCFeature;
     private Feature docStoreEmbeddedVerificationFeature;
+    private Feature docStoreFullGCModeGapOrphansFeature;
+    private Feature docStoreFullGCModeEmptyPropertiesFeature;
```
I think we aimed to remove the features here?

shodaaan replied:
Fixed, thank you for the comment.

stefan-egli commented on oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/VersionGCTest.java:
```diff
@@ -531,6 +531,57 @@ public void testResetFullGCDryRunMode() throws Exception {
+                fullGcModeNotAllowedValue);
+
+        assertEquals("Starting VersionGarbageCollector with not applicable / not allowed value" +
+                "will set fullGcMode to default NONE", gc.getFullGcMode(), VersionGarbageCollector.FullGCMode.NONE);
```
```suggestion
                "will set fullGcMode to default NONE", VersionGarbageCollector.getFullGcMode(), VersionGarbageCollector.FullGCMode.NONE);
```

static method should be accessed in a static way

shodaaan replied:
Fixed the comment, unit tests and static change, thank you for the suggestions

stefan-egli reviewed (commented):
I think the unrelated changes in particular in DocumentNodeStore should be avoided. Also, the feature code is not fully removed yet

stefan-egli commented:
PS: I think we should adjust jira/pr titles to talk about osgi config instead of feature toggle (to avoid future confusion)

shodaaan committed:
- made some of the requested PR changes
- reverted CommitBuilderTest to previous version

rishabhdaim commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java:
```diff
@@ -47,13 +47,13 @@
 import org.apache.jackrabbit.oak.plugins.document.util.MongoConnection;
 import org.apache.jackrabbit.oak.run.commons.Command;
-import org.apache.jackrabbit.oak.plugins.document.VersionGCOptions;
-import org.apache.jackrabbit.oak.plugins.document.VersionGarbageCollector;
```
please revert these import changes.

shodaaan replied:
done

rishabhdaim commented on oak-run/src/main/java/org/apache/jackrabbit/oak/run/RevisionsCommand.java:
```diff
@@ -69,10 +69,11 @@
+import static org.apache.jackrabbit.oak.plugins.document.util.Utils.getFullGCMode;
 import static org.apache.jackrabbit.oak.plugins.document.util.Utils.getIdFromPath;
 import static org.apache.jackrabbit.oak.plugins.document.util.Utils.getRootDocument;
-import static org.apache.jackrabbit.oak.plugins.document.util.Utils.isFullGCEnabled;
```
same as above.

rishabhdaim commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java:
```diff
@@ -211,6 +211,29 @@ static FullGCMode getFullGcMode() {
+     * if it is set to any other values than the supported ones.
+     * @param fullGcModeConfig
+     */
+    static void setFullGcMode(int fullGcModeConfig) {
```
I would suggest renaming variable name to `fullGcMode`

shodaaan replied:
Done.

rishabhdaim commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/util/Utils.java:
```diff
@@ -1009,6 +1009,16 @@ public static boolean isEmbeddedVerificationEnabled(final DocumentNodeStoreBuild
+     * @param builder instance for DocumentNodeStoreBuilder
+     * @return true if full GC is enabled else false
+     */
+    public static int getFullGCMode(final DocumentNodeStoreBuilder<?> builder) {
```
This utility method doesn't make sense, it only returns `fullGcMode` from the `builder` object. I would remove this.

shodaaan replied:
Fixed.

rishabhdaim commented on oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/util/UtilsTest.java:
```diff
@@ -236,6 +237,30 @@ public void fullGCDisabledForRDB() {
+    @Test
+    public void fullGCModeDefaultValue() {
+        int fullGCModeDefaultValue = getFullGCMode(newDocumentNodeStoreBuilder());
+        final int FULL_GC_MODE_NONE = 0;
```
please use camelCase here.

shodaaan replied:
Done.

rishabhdaim commented on oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/util/UtilsTest.java:
```diff
@@ -236,6 +237,30 @@ public void fullGCDisabledForRDB() {
+    @Test
+    public void fullGCModeSetViaConfiguration() {
+        DocumentNodeStoreBuilder<?> builder = newDocumentNodeStoreBuilder();
+        final int FULL_GC_MODE_GAP_ORPHANS = 2;
```
same as above.

shodaaan replied:
Done.

shodaaan committed:
- reverted Utils.[static method] changes made in DocumentNodeStore

shodaaan committed:
- re-added spaces between switch case value and ":" - against suggested Java formatting but consistent with how code was originally written

shodaaan committed:
- removed Utils static method getFullGCMode and adjusted rest of files as necessary
- made other requested changes

stefan-egli commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java:
```diff
@@ -324,4 +325,14 @@
+            description = "Integer value indicating which Full GC mode should be enabled for " +
+                    "document node store. The Default value is " + DEFAULT_FULL_GC_MODE +
+                    ". Note that this value can be overridden via framework " +
+                    "property 'oak.documentstore.fullGCMode'." +
+                    "FullGC can be entirely enabled / disabled with the variable fullGCEnabled, unless fullGCEnabled" +
+                    "is set to true, the fullGCMode will be ignored.")
```
```suggestion
                    "property 'oak.documentstore.fullGCMode'. " +
                    "FullGC can be entirely enabled / disabled with the variable fullGCEnabled, unless fullGCEnabled " +
                    "is set to true, the fullGCMode will be ignored.")
```

stefan-egli reviewed (commented):
an osgi config formatting detail added - plus the conflicts with trunk would need to be resolved - otherwise looking good

shodaaan committed:
Merge branch 'trunk' into OAK-10896

shodaaan committed:
- added missing spaces in fullGCMode variable description

rishabhdaim commented on oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoDocumentNodeStoreBuilderTest.java:
```diff
@@ -111,4 +110,11 @@ public void collectionCompressionDisabled() {
+    @Test
+    public void fullGCModeDefaultValue() {
+        MongoDocumentNodeStoreBuilder builder = new MongoDocumentNodeStoreBuilder();
+        final int FULL_GC_MODE_NONE = 0;
```
please use camelCase here.

shodaaan committed:
- fixed camel case

stefan-egli reviewed (approved):
+1 (there's some unrelated newlines, but let's leave them for now)

rishabhdaim commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/rdb/RDBDocumentNodeStoreBuilder.java:
```diff
@@ -154,6 +157,18 @@ public RDBDocumentNodeStoreBuilder setFullGCExcludePaths(@NotNull String[] exclu
+    @Override
+    public RDBDocumentNodeStoreBuilder setFullGCMode(int v) {
+        // fullGC modes are not supported for RDB
+        log.warn("FullGC modes are not supported for RDB");
```
this should return the `builder` object.

shodaaan replied:
Fixed.

shodaaan committed:
- fixed compilation error due to incorrect merge

rishabhdaim commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/rdb/RDBDocumentNodeStoreBuilder.java:
```diff
@@ -143,7 +147,6 @@ public RDBDocumentNodeStoreBuilder setFullGCIncludePaths(@NotNull String[] inclu
         return thisBuilder();
     }
 
-    @Override
```
please revert this.

shodaaan committed:
- re-added @Override that was missed in merge

rishabhdaim committed:
Update Configuration.java

rishabhdaim committed:
Update DocumentNodeStore.java

rishabhdaim committed:
Update VersionGarbageCollector.java

rishabhdaim committed:
Update VersionGarbageCollector.java

rishabhdaim committed:
Update MongoDocumentNodeStoreBuilderTest.java

stefan-egli commented:
+1, I'll do a squash-n-merge next. was wondering why `DocumentNodeStoreConfigTest` fails - but that's unrelated, fails on trunk too, so this is good to go ahead.
